prometheus/go_collector.go

```diff
@@ -261,14 +157,28 @@ func (c *baseGoCollector) Collect(ch chan<- Metric) {
 	ch <- MustNewConstMetric(c.goInfoDesc, GaugeValue, 1)
 }
 
-func memstatNamespace(s string) string {
-	return "go_memstats_" + s
+// memStatsMetric provide description, evaluator, runtime/metrics name, and
+// value type for memstat metric.
+type memStatsMetric struct {
+	last   float64
+	eval   func(*runtime.MemStats) float64
+	metric Metric
+}
+
+func (m *memStatsMetric) update(memStats *runtime.MemStats) Metric {
```

Comment:
This is not concurrency-safe. Metrics collection can happen concurrently, and then all kind of weirdness will happen here.